Remove Typescript global dep

littleninja opened the pull request:
Response to #608. 

- **Add Typescript as a dev dependency** Main revision. Add `typescript` to dev dependencies and move `@types` there, too.
- **Add post-install script** Add a script to run after `npm install` to compile Typescript. 
- **Add to .npmignore to reduce package size.** `npm pack` run locally on master produces a tarball at about ~737KB. With the inclusions to `.npmignore` (source code, tests, project configuration) the size is ~389KB or nearly half the size.
- **Update docs and `appveyor.yml`** Remove references to global install and manually running the compile step.
- **Tidy `tsconfig.json`**

littleninja commented on .npmignore:
```diff
@@ -1 +1,17 @@
+
+# Project configuration
+.appveyor.yml
+.eslintrc.json
```
Is `.eslintrc.json` a project configuration or is it part of the package? If part of the package, my mistake and I'll remove it.

prettydiff replied:
I do not understand the distinction.  They are both configuration files.  The appveyor file defines the build steps for the appveyor Windows continuous integration service.

The eslintrc file defines the configuration for executing ESLint on the project.  This where I define my opinions and subjective thoughts on code style and I refuse to set or define any other subjectivity.  I am not a believer in code style or enforcing subjective opinions on people, but consistency is important.

littleninja replied:
Oops, I should clarify: I'm asking whether or not the `.eslintrc.json` file should be included in the published module. My guess was project configuration for those working on the source code, but I wanted to be sure given the global dependency on eslint. Totally support helping contributors keep a consistent style in your codebase 💯

prettydiff replied:
The .eslintrc.json file is still helpful in the NPM package, because the *test* and *lint* commands will continue to execute if ESLint is installed on the user's system.  The other configuration files don't matter to the NPM package though.

littleninja commented:
At least one test is failing. I also found where `services.js` will use `tsc` to build sparser, I need to look around a little more. I don't see an option to change this into a draft, so I'm going to close and I'll reopen a draft pull request. This needs a little more work.

prettydiff commented:
Thank you for your work on this.  Just let me know if you need another set of eyes.

prettydiff commented:
I just looked at the failed appveyor test.  Before giving up on this it looks like the difference could be related to some code changes I have made over the past month.  Do a git pull from master and merge into your branch.  I bet this will fix the failing tests.

littleninja commented:
Thanks! Not giving up, just didn't want this to be merged when it's not ready. Should I reopen this pull request if I'm still working on it?

I fixed most of the failing tests by create a test copy of the `tsconfig.json` file. For a couple `hash` tests that exclude the tests directory, I updated the hash value.

There is still `services.ts` to attend to, the `sparser()` and `typescript()` functions. I haven't looked closely yet to know what direction to take.

Thanks again for taking a look at this!

littleninja commented:
Looking into `services.ts` and the commands for `build` and `test` has been great for context. Are `prettydiff build` and `prettydiff test` for development only? If development only, do you mind if I attempt a refactor? I think they could be moved to their own `build/*.ts` files. If you don't mind me taking a swing at it, I'd like to hold on this pull request until build/test commands are separated from service commands.

prettydiff commented:
Yes, mostly.  The build command is mostly for dev, but is required if the application is cloned from github because I do not commit the JavaScript files.  Pulling the application off NPM contains the JavaScript files so in that case the build command isn't needed at all, unless a user wishes to play around and rebuild the application.

The test command, which is composed of various child commands is really only for development, but can be run at any time as a status or health indicator.

I appreciate a refactor.  I am interested in your ideas and seeing your work.  Please keep in mind all this architecture is so far just me.  I really appreciate a second opinion that challenges my assumptions.  Your effort is a huge help.

prettydiff commented:
I resolved some merge conflicts in the simulations.ts file and the package.json file.  Don't worry about the simulations.ts file because that conflict had nothing to do with your pull request.  Should the `@types/node` dependency be removed from the package.json file?  I am not sure that was a goal.

littleninja commented:
Thanks for resolving merge conflicts. I'm making assumptions about the architecture and its future, and I'm hoping you'll check me if I go too far.

My goal is to separate pulling and building the source (TS) from installing the package (JS). I think it's possible to write build scripts that produce the JavaScript files as an artifact and publish this. It's proving a bigger change and a bit tricky but I'm up for the challenge. If there's no rush, I'm happy to continue plucking away at this, I may give it a fresh approach on a new branch.

prettydiff commented:
Absolutely, and there is definitely no rush.  I appreciate the help.

prettydiff commented on package.json:
```diff
@@ -7,11 +7,14 @@
     "devDependencies": {
+        "@types/node": "^10.14.13",
         "ace-builds": "^1.4.5",
+        "typescript": "^3.5.2",
```
What is the benefit for adding TypeScript both as a regular dependency and a dev dependency?

prettydiff commented on tsconfig.json:
```diff
@@ -5,14 +5,10 @@
-        "3",
         "js",
-        "ignore",
         "node_modules"
```
Those extra lines for TS exclusion are there because I have directories on my website at https;//prettydiff.com that I don't want TypeScript looking into.
